From b831a6b7ec34509c9adcb72a473928a52f8fca72 Mon Sep 17 00:00:00 2001
From: Flopster101 <nahuelgomez329@gmail.com>
Date: Fri, 15 May 2026 23:43:48 -0300
Subject: [PATCH 4/4] minuitwrp: graphics: Fix screenshot color swap for Exynos
 fbdev

This fixes blue-tinted screenshots on Exynos devices by bypassing the
redundant Red/Blue swap when TW_EXYNOS_FBDEV_COLOR_SWAP is active.

Additionally, it improves the utility's reliability by using direct
framebuffer memory copy instead of Pixelflinger, which avoids potential
format mismatch issues and crashes on some hardware.

Signed-off-by: Flopster101 <nahuelgomez329@gmail.com>
---
 minuitwrp/graphics_utils.cpp | 111 ++++++++++++-----------------------
 1 file changed, 36 insertions(+), 75 deletions(-)

diff --git a/minuitwrp/graphics_utils.cpp b/minuitwrp/graphics_utils.cpp
index 9860d89f..60b1acc5 100644
--- a/minuitwrp/graphics_utils.cpp
+++ b/minuitwrp/graphics_utils.cpp
@@ -34,96 +34,57 @@ extern unsigned int gr_rotation;
 int gr_save_screenshot(const char *dest)
 {
     uint32_t y, stride_bytes;
-    volatile int res = -1;
-    GGLContext *gl = NULL;
-    GGLSurface surface;
-    uint8_t * volatile img_data = NULL;
-    uint8_t *ptr;
-    FILE * volatile fp = NULL;
+    int res = -1;
+    uint8_t *img_data = NULL, *row_buf = NULL;
+    FILE *fp = NULL;
     png_structp png_ptr = NULL;
     png_infop info_ptr = NULL;
 
-    fp = fopen(dest, "wb");
-    if(!fp)
-        goto exit;
+    if (!(fp = fopen(dest, "wb"))) return -1;
 
-    img_data = (uint8_t *)malloc(gr_mem_surface.stride * gr_mem_surface.height * gr_draw->pixel_bytes);
-    if (!img_data) {
-        printf("gr_save_screenshot failed to malloc img_data\n");
-        goto exit;
-    }
-    surface.version = sizeof(surface);
-    surface.width = gr_mem_surface.width;
-    surface.height = gr_mem_surface.height;
-    surface.stride = gr_mem_surface.stride;
-    surface.data = img_data;
-
-#if defined(RECOVERY_BGRA)
-    surface.format = GGL_PIXEL_FORMAT_BGRA_8888;
-#else
-    surface.format = GGL_PIXEL_FORMAT_RGBA_8888;
-#endif
-
-    gglInit(&gl);
-    gl->colorBuffer(gl, &surface);
-    gl->activeTexture(gl, 0);
-
-    if(gr_mem_surface.format == GGL_PIXEL_FORMAT_RGBX_8888)
-        gl->disable(gl, GGL_BLEND);
-
-    gl->bindTexture(gl, &gr_mem_surface);
-    gl->texEnvi(gl, GGL_TEXTURE_ENV, GGL_TEXTURE_ENV_MODE, GGL_REPLACE);
-    gl->texGeni(gl, GGL_S, GGL_TEXTURE_GEN_MODE, GGL_ONE_TO_ONE);
-    gl->texGeni(gl, GGL_T, GGL_TEXTURE_GEN_MODE, GGL_ONE_TO_ONE);
-    gl->enable(gl, GGL_TEXTURE_2D);
-    gl->texCoord2i(gl, 0, 0);
-    gl->recti(gl, 0, 0, gr_mem_surface.width, gr_mem_surface.height);
-
-    gglUninit(gl);
-    gl = NULL;
+    stride_bytes = gr_draw->row_bytes;
+    img_data = (uint8_t *)malloc(gr_draw->height * stride_bytes);
+    if (!img_data) goto exit;
+    memcpy(img_data, gr_draw->data, gr_draw->height * stride_bytes);
 
-    png_ptr = png_create_write_struct(PNG_LIBPNG_VER_STRING, NULL, NULL, NULL);
-    if (!png_ptr)
-        goto exit;
-
-    info_ptr = png_create_info_struct(png_ptr);
-    if (info_ptr == NULL)
-        goto exit;
-
-    if (setjmp(png_jmpbuf(png_ptr)))
-        goto exit;
+    if (!(png_ptr = png_create_write_struct(PNG_LIBPNG_VER_STRING, NULL, NULL, NULL))) goto exit;
+    if (!(info_ptr = png_create_info_struct(png_ptr))) goto exit;
+    if (setjmp(png_jmpbuf(png_ptr))) goto exit;
 
     png_init_io(png_ptr, fp);
-    png_set_IHDR(png_ptr, info_ptr, surface.width, surface.height,
-         8, PNG_COLOR_TYPE_RGB, PNG_INTERLACE_NONE,
-         PNG_COMPRESSION_TYPE_BASE, PNG_FILTER_TYPE_BASE);
+    png_set_IHDR(png_ptr, info_ptr, gr_draw->width, gr_draw->height, 8, 
+                 PNG_COLOR_TYPE_RGB, PNG_INTERLACE_NONE,
+                 PNG_COMPRESSION_TYPE_BASE, PNG_FILTER_TYPE_BASE);
     png_write_info(png_ptr, info_ptr);
 
-    // To remove the alpha channel for PNG_COLOR_TYPE_RGB format,
-    png_set_filler(png_ptr, 0, PNG_FILLER_AFTER);
+    row_buf = (uint8_t*)malloc(gr_draw->width * 3);
+    if (!row_buf) goto exit;
 
-    ptr = img_data;
-    stride_bytes = surface.stride*4;
-    for(y = 0; y < surface.height; ++y)
-    {
-        png_write_row(png_ptr, ptr);
-        ptr += stride_bytes;
+    for(y = 0; y < gr_draw->height; ++y) {
+        uint8_t *src = img_data + y * stride_bytes, *dst = row_buf;
+        for(unsigned int x = 0; x < gr_draw->width; ++x, src += 4, dst += 3) {
+#if defined(EXYNOS_FBDEV_COLOR_SWAP)
+            dst[0] = src[0]; dst[1] = src[1]; dst[2] = src[2];
+#else
+            if (gr_draw->format == GGL_PIXEL_FORMAT_BGRA_8888) {
+                dst[0] = src[2]; dst[1] = src[1]; dst[2] = src[0];
+            } else {
+                dst[0] = src[0]; dst[1] = src[1]; dst[2] = src[2];
+            }
+#endif
+        }
+        png_write_row(png_ptr, row_buf);
     }
 
     png_write_end(png_ptr, NULL);
-
     res = 0;
+
 exit:
-    if(info_ptr)
-        png_free_data(png_ptr, info_ptr, PNG_FREE_ALL, -1);
-    if(png_ptr)
-        png_destroy_write_struct(&png_ptr, (png_infopp)NULL);
-    if(gl)
-        gglUninit(gl);
-    if(img_data)
-        free(img_data);
-    if(fp)
-        fclose(fp);
+    if(row_buf) free(row_buf);
+    if(info_ptr) png_free_data(png_ptr, info_ptr, PNG_FREE_ALL, -1);
+    if(png_ptr) png_destroy_write_struct(&png_ptr, NULL);
+    if(img_data) free(img_data);
+    if(fp) fclose(fp);
     return res;
 }
 
-- 
2.54.0

